orttraining/orttraining/core/session/training_session.cc

```diff
@@ -916,11 +1040,211 @@ common::Status TrainingSession::Run(const RunOptions& run_options, IOBinding& io
+  };
+
+  int id = -1;
+
+  // Forward Recv
+  id = !training_mode ? -1 : pipeline_schedule_.GetForwardRecvWaitedEvent(batch_id, stage_id);
+  append_to_io_binding(pipeline_context_.pipeline_tensor_names.forward_recv_waited_event_name, id);
+  id = !training_mode ? -1 : pipeline_schedule_.GetForwardRecvRecordedEvent(batch_id, stage_id);
+  append_to_io_binding(pipeline_context_.pipeline_tensor_names.forward_recv_recorded_event_name, id);
+
+  // Forward Send
+  id = !training_mode ? -1 : pipeline_schedule_.GetForwardSendWaitedEvent(batch_id, stage_id);
+  append_to_io_binding(pipeline_context_.pipeline_tensor_names.forward_send_waited_event_name, id);
+  id = !training_mode ? -1 : pipeline_schedule_.GetForwardSendRecordedEvent(batch_id, stage_id);
+  append_to_io_binding(pipeline_context_.pipeline_tensor_names.forward_send_recorded_event_name, id);
+
+  // Backward Recv
+  id = !training_mode ? -1 : pipeline_schedule_.GetBackwardRecvWaitedEvent(batch_id, stage_id);
+  append_to_io_binding(pipeline_context_.pipeline_tensor_names.backward_recv_waited_event_name, id);
+  id = !training_mode ? -1 : pipeline_schedule_.GetBackwardRecvRecordedEvent(batch_id, stage_id);
+  append_to_io_binding(pipeline_context_.pipeline_tensor_names.backward_recv_recorded_event_name, id);
+
+  // Backward Send
+  id = !training_mode ? -1 : pipeline_schedule_.GetBackwardSendWaitedEvent(batch_id, stage_id);
+  append_to_io_binding(pipeline_context_.pipeline_tensor_names.backward_send_waited_event_name, id);
+  id = !training_mode ? -1 : pipeline_schedule_.GetBackwardSendRecordedEvent(batch_id, stage_id);
+  append_to_io_binding(pipeline_context_.pipeline_tensor_names.backward_send_recorded_event_name, id);
+
+  // Forward Compute
+  id = !training_mode ? -1 : pipeline_schedule_.GetForwardComputeWaitedEvent(batch_id, stage_id);
+  append_to_io_binding(pipeline_context_.pipeline_tensor_names.forward_compute_waited_event_name, id);
+  id = !training_mode ? -1 : pipeline_schedule_.GetForwardComputeRecordedEvent(batch_id, stage_id);
+  append_to_io_binding(pipeline_context_.pipeline_tensor_names.forward_compute_recorded_event_name, id);
+
+  // Backward Compute
+  id = !training_mode ? -1 : pipeline_schedule_.GetBackwardComputeWaitedEvent(batch_id, stage_id);
+  append_to_io_binding(pipeline_context_.pipeline_tensor_names.backward_compute_waited_event_name, id);
+  id = !training_mode ? -1 : pipeline_schedule_.GetBackwardComputeRecordedEvent(batch_id, stage_id);
+  append_to_io_binding(pipeline_context_.pipeline_tensor_names.backward_compute_recorded_event_name, id);
+}
+
+// This function splits input batch into several sub-batches and then
+// run those sub-batches using pipeline parallel. This function
+// is responsible for adding pipeline-related feeds such as event IDs before
+// calling the graph.
+common::Status TrainingSession::RunWithPipeline(const RunOptions& run_options, IOBinding& io_binding) {
+  const size_t num_steps = pipeline_context_.num_pipeline_steps;
```

Comment:
Is it clearer to rename -> num_pipeline_stages (because we're really talking about the structure of the pipeline

Reply:
`num_pipeline_steps` and `num_pipeline_stages` are different. `num_pipeline_steps` is the number of sub-batches. `num_pipeline_stages` is the number of model partitions.

Reply:
Could we say something more descriptive then like num_pipeline_ga_steps, or num_pipeline_microbatches?

Reply:
Sure. Will use `num_pipeline_micro_batches`.

---
In reply to: [532596194](https://github.com/microsoft/onnxruntime/pull/5815#discussion_r532596194) [](ancestors = 532596194)